### PR TITLE
Feature/add dmrl: Add DMRL Model

### PR DESCRIPTION
**Pull request by mabeckers**
### Description
I added Disentangled Multimodal Representation Learning (https://arxiv.org/pdf/2203.05406.pdf) as the DMRL model to cornac.
In the context of this addition I had to:
1. add the PWLearningSampler: A Sampler to be wrapped around PyTorche DataLoader class that does the necessary loading of the data in a way that DMRL training mechanism requires it (pairwise based ranking approach)
2. add the TransformersTextModality (as described in the paper) to encode textual features into latent space
3. add the TransformersVisionModality (as described in the paper) to encode visual features into latent space
4. add the DistanceCorrelationCalculator which can be use to calculate the disentangled loss (as described in the paper)
5. add dmrl and recom_dmrl as described by paper and cornac framework
6. added tests for all above described modules
7. had to modify BaseMethod to include TransformersTextModality as allowed testmodality
8. TransformersVision Modality not yet included in BaseMehthod as not used in dmrl examples (preencoded cornac vision features were used). Please add if wanted.

### Checklist:

- [x] I have added tests.
- [x] I have updated the documentation accordingly.
- [x] I have updated `README.md` (if you are adding a new model).
- [x] I have updated `examples/README.md` (if you are adding a new example).
- [x] I have updated `datasets/README.md` (if you are adding a new dataset).

### CODE REVIEW

**Comment by tqtg** [comment]
Hi @mabeckers, thanks for the contribution. It's great to see DMRL being added into Cornac. However, there are a few things that we might need to reconsider. First, each model in Cornac is very self-contained and model dependencies should not be added as global requirements. This is to minimize the maintenance effort for the core functions, also to facilitate a wider-range of model implementation. With that said, there are two directions to proceed with the DMRL model:
1. Considering the model taking in raw text and raw video. With this approach, we should bring the text/video transformer-based encoders as part of the model implementation. For example, [CDL](https://github.com/PreferredAI/cornac/tree/master/cornac/models/cdl) model has an autoencoder for text.
2. Considering the model taking in text/video features. In this case, we should do text/video encoding separately (possibly as part of the example) prior to model training in Cornac. With the embedding ready, we simply employ the `FeatureModality` for either text/video. We can consider if additional `VideoModality` is needed or `ImageModality` could be used as an alternative for data input.

Hope that my explanation is clear enough. Happy to chat more.

**Comment by mabeckers** [comment]
Hi @tqtg. thanks for taking a look at my PR. Yeah I notice that the two new transformer modalities introduced more general dependencies. I can go ahead and move those modules inside of recom_dmrl, that way the model receives basic text and image as input. I will make it general so that in case one already has encoded features from somewhere (say it comes with the example) the model can take that in as well and will not run another layer of encoding on top of that feature set.
Does that sound good to you?

Thanks,
Max

**Comment by tqtg** [comment]
sounds good to me. Let's do that and see how it goes.

File: cornac/eval_methods/base_method.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 from scipy.sparse import csr_matrix
 from tqdm.auto import tqdm
 
+
```
Comment (tqtg) [inline cornac/eval_methods/base_method.py:23]: Let's remove this change

File: cornac/models/__init__.py [hunk 2]
```diff
@@ -81,3 +81,12 @@
 from .vbpr import VBPR
 from .vmf import VMF
 from .wmf import WMF
+from .dmrl import recom_dmrl
```
Comment (tqtg) [inline cornac/models/__init__.py:84]: As a convention, we import the model class name directly instead of the module. Also, please place the import statement in the correct alphabetical order.

File: cornac/models/__init__.py [hunk 3]
```diff
@@ -81,3 +81,12 @@
...
+
+try:
```
Comment (tqtg) [inline cornac/models/__init__.py:86]: Please remove this change, it's outdated

File: cornac/models/dmrl/pwlearning_sampler.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        list_of_indexs: List[int]
+            List of indexs to sample from the uir_tuple given in cornac dataset
+        """
+        batch_size = len(list_of_indexs)
+        users = self.user_array[list_of_indexs]
+        pos_items = self.item_array[list_of_indexs]
+
+        pos_u_i = np.vstack([users, pos_items]).T
+
+        # sample negative items per user
```
Comment (tqtg) [inline cornac/models/dmrl/pwlearning_sampler.py:64]: This is similar to the [Dataset.uij_iter()](https://github.com/PreferredAI/cornac/blob/c6035980a3ee0735261f8327b9c3a1da870f7de8/cornac/data/dataset.py#L490) which I believe it could be reused for simpler implementation.

File: cornac/models/dmrl/pwlearning_sampler.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        # first select index tuple
+        user = self.user_array[index]
+        item = self.item_array[index]
+
+        # now construct positive case
+        pos_u_i = [user, item]
+
+        i = 0
+        neg_i = []
+        while i < self.num_neg:
```
Comment (tqtg) [inline cornac/models/dmrl/pwlearning_sampler.py:100]: https://github.com/PreferredAI/cornac/blob/c6035980a3ee0735261f8327b9c3a1da870f7de8/cornac/data/dataset.py#L490

File: cornac/models/dmrl/recom_dmrl.py [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+
+from typing import List, Tuple
+
+import numpy as np
+import torch
```
Comment (tqtg) [inline cornac/models/dmrl/recom_dmrl.py:19]: We don't have torch, PIL, transformers installed by default. Therefore, it will cause import errors when others not using DMRL model. Please move the import statements into `_fit_dmrl()` and `score()` functions accordingly.

File: README.md [hunk 7]
```diff
@@ -147,6 +147,7 @@ The recommender models supported by Cornac are listed below. Why don't you join
 
 | Year | Model and paper | Model type | Require-ments | Examples |
 | :---: | --- | :---: | :---: | :---: |
+| 2022 | [Disentangled Multimodal Representation Learning for Recommendation)](cornac/models/bivaecf), [paper](https://arxiv.org/pdf/2203.05406.pdf) | Collaborative Filtering / Content-Based | [reqs](cornac/models/dmrl/requirements.txt) | [exp](examples/dmrl_example.py)
```
Comment (tqtg) [inline README.md:150]: The link refers to bivaecf

File: requirements.txt [hunk 8]
```diff
@@ -2,4 +2,8 @@ numpy
 scipy
 Cython
 tqdm
-powerlaw
\ No newline at end of file
+powerlaw
+sentence_transformers
```
Comment (tqtg) [inline requirements.txt:6]: We should remove the model dependencies. We're not going to install them by default.

**Review by tqtg — CHANGES_REQUESTED** [review]

File: cornac/models/dmrl/pwlearning_sampler.py (hunk 4 above)
Comment (mabeckers) [reply cornac/models/dmrl/pwlearning_sampler.py:64]: Not exactly what's needed here as uij_iter() returns a set of users, 1 positive and 1 negative array. For DMRL we need a set of 1 user array, 1 pos item array and num_neg (e.g. 4) negative items per user in a 2D array.
But thanks for the hint! I did take a look at it and it is indeed not far off.

**Comment by mabeckers** [comment]
Made the requested changes, please let me know if there's anything else I can change for this PR. Also remerged with the latest cornac master.
Thanks!

File: cornac/models/dmrl/pwlearning_sampler.py (hunk 4 above)
Comment (tqtg) [reply cornac/models/dmrl/pwlearning_sampler.py:64]: Got it. I overlooked the number of negative samples

**Review by tqtg — APPROVED** [review]

**Comment by tqtg** [comment]
@mabeckers I did some changes to make the tests work and also refactoring. Please have a look and see if they make sense to you.

**Comment by mabeckers** [comment]
Everything looks great to me!

**Comment by tqtg** [comment]
Hey @mabeckers, there is something that we need to modify about the model input (text and image modalities). By design, we don't input the modalities directly to the model, but we input them to an evaluation method (e.g., RatioSplit). The reason is that the modalities will be aligned with user/item data splitting and user/item ID being mapped properly. Taking CDL model as an example, we input text modality to the RatioSplit eval method ([here](https://github.com/mabeckers/cornac/blob/a29b9fafb385cd69555a484dc4e3a0720eab465f/examples/cdl_example.py#L45)) and we can access the text modality inside the model implementation via the train_set ([here](https://github.com/PreferredAI/cornac/blob/a2ee37e3919564a38d308507b564de5c5856782d/cornac/models/cdl/recom_cdl.py#L189)). Can we work on this last change before we merge the model into Cornac?

**Comment by mabeckers** [comment]
Hey @tqtg Yeah I understand that's how the cornac framework works with modalities, which is why until commit [9fc96b3](https://github.com/PreferredAI/cornac/pull/597/commits/9fc96b3abc68e3f1583508d957bd8895b7b44e33) I had it that way and was feeding modalities from the outside to the RatioSplit Instance. I only changed it and moved them inside the model because you mentioned you didn't want to add any general dependencies (such as new TransformerModalities) to the cornac core but move that into the DMRL folder and have the model accept raw text and images. That's why I moved it out of RatioSplit. I am happy to reverse the commit back to the earlier version and introduce TransformerVisionModadality and TransformerTextModality as new general modality encoders. I can of course also just keep them in the DMRL folder and still use them as normal modalities and input to the RatioSplit instance. Just let me know which way you would prefer it.
Thanks!

**Comment by tqtg** [comment]
My point is that you can reuse TextModality and ImageModality to hold the image/text corpus and input them into RatioSplit to perform data splitting. The only part we want to move inside model implementation is where we use Transformers to encode the raw data. Does that make sense to you?

**Comment by mabeckers** [comment]
Ok I see. So if I am understanding you correctly you would want the example file running the DMRL example to look something like this?:

"""Example for Disentangled Multimodal Recommendation, with only feedback and textual modality.
For an example including image modality please see dmrl_clothes_example.py"""

import cornac
from cornac.data import Reader
from cornac.datasets import citeulike
from cornac.eval_methods import RatioSplit
from cornac.models.dmrl.recom_dmrl import TextModalityInput

# The necessary data can be loaded as follows
docs, item_id_ordering_text = citeulike.load_text()
feedback = citeulike.load_feedback(reader=Reader(item_set=item_id_ordering_text))



text_modality_input = TextModalityInput(item_id_ordering_text, docs)

# Instantiate DMRL recommender
dmrl_recommender = cornac.models.dmrl.DMRL(
    batch_size=4096,
    epochs=20,
    log_metrics=False,
    learning_rate=0.01,
    num_factors=2,
    decay_r=0.5,
    decay_c=0.01,   
    num_neg=3,
    embedding_dim=100,
    text_features=text_modality_input)

### NEW METHOD THAT HOLDS THE TRANSFORMER ENCODING WITHIN DMRL MODEL:
item_text_modality = dmrl_recommender.encode_text() # returns a generic feature modality (or even a TextModality) # where pre-encoded text is given in .features attribute and uses Transformer internally.

# Define an evaluation method to split feedback into train and test sets
ratio_split = RatioSplit(
    data=feedback,
    test_size=0.2,
    exclude_unknowns=True,
    verbose=True,
    seed=123,
    rating_threshold=0.5,
    item_text = item_text_modality
)

# Use Recall@300 for evaluations
rec_300 = cornac.metrics.Recall(k=300)
prec_30 = cornac.metrics.Precision(k=30)

# Put everything together into an experiment and run it
cornac.Experiment(eval_method=ratio_split, models=[dmrl_recommender], metrics=[prec_30, rec_300]).run()

**Comment by tqtg** [comment]
@mabeckers I made some changes to illustrate my idea. Please have a look and let me know if they make sense to you. We can further refactor the code to remove some unused parts.

**Comment by mabeckers** [comment]
Had to set preencode=True (that means to be pre-encoded as part of TransformersModality init, preencoded means ists already pre-encoded from outside), but other than that looks very good! I understand now what you meant. We use TextModality for data splitting and id mapping on outside and "overwrite" it on the inside with TransformerModalities. Just running some final checks then will commit! Thanks for showing me this way of doing it. Only downside here is that we call vectorizer.fit_transform(self.corpus) in _build_text() of the TextModality when all we want is _swap_text() ... so a little overhead but I'm fine doing it that way :)

**Comment by tqtg** [comment]
@mabeckers OK, I though it should be encoded batch by batch during training thus `preencode=False`. Anw, please help check because I might misinterpret your implementation.

For the basic text transformation overhead, I'm aware of that and it might be an issue with a big text corpus. I was thinking of using tokenizer as the indicator whether we want to do any transformation or not. If the tokenizer is not provided during the initialization of the TextModality, we just bypass the `_build_text()` call. What do you think about this?

**Comment by mabeckers** [comment]
yeah the tokenizer is a good idea. That would make sense. The TransformerModalities work both in batch as well as pre-encoding. Just using them as pre-encoding in my examples bc it makes runtime a lot faster. I just pushed my final changes and checks. Thanks

**Comment by tqtg** [comment]
> yeah the tokenizer is a good idea. That would make sense. The TransformerModalities work both in batch as well as pre-encoding. Just using them as pre-encoding in my examples bc it makes runtime a lot faster. I just pushed my final changes and checks. Thanks

Cool! This PR looks good to me. Let's merge it and have another one to update the TextModality. Thanks @mabeckers!
